File: maze.cpp
// maze.cpp
#include <iostream>
#include <vector>
#include <random>
#include <chrono>
#include <thread>
#include <string>
#include <fstream>
#include <json/json.h>
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>

using namespace std;

const string RESET = "\033[0m";
const string RED = "\033[91m";
const string GREEN = "\033[92m";
const string YELLOW = "\033[93m";
const string BLUE = "\033[94m";
const string CYAN = "\033[96m";
const string GRAY = "\033[90m";
const string BOLD = "\033[1m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

const int DEFAULT_SIZE = 10;
const int DEFAULT_COINS = 5;

string getHomeDir() {
    const char* home = getenv("HOME");
    if (!home) home = getenv("USERPROFILE");
    return string(home);
}

Json::Value loadRecords() {
    ifstream f(getHomeDir() + "/.maze_records.json");
    Json::Value root;
    if (!f) return root;
    f >> root;
    return root;
}

void saveRecords(const Json::Value& records) {
    ofstream f(getHomeDir() + "/.maze_records.json");
    f << records.toStyledString();
}

void clearScreen() {
    cout << "\033[2J\033[1;1H";
}

vector<vector<int>> generateMaze(int size) {
    vector<vector<int>> maze(size, vector<int>(size, 1));
    random_device rd;
    mt19937 g(rd());

    function<void(int,int)> dfs = [&](int x, int y) {
        maze[x][y] = 0;
        vector<pair<int,int>> dirs = {{0,2},{0,-2},{2,0},{-2,0}};
        shuffle(dirs.begin(), dirs.end(), g);
        for (auto& d : dirs) {
            int nx = x + d.first, ny = y + d.second;
            if (nx > 0 && nx < size-1 && ny > 0 && ny < size-1 && maze[nx][ny] == 1) {
                maze[x + d.first/2][y + d.second/2] = 0;
                dfs(nx, ny);
            }
        }
    };
    dfs(1, 1);
    maze[size-2][size-1] = 0;
    maze[size-1][size-1] = 0;
    return maze;
}

vector<pair<int,int>> placeCoins(const vector<vector<int>>& maze, int numCoins, int size) {
    vector<pair<int,int>> coins;
    random_device rd;
    mt19937 g(rd());
    uniform_int_distribution<int> dist(1, size-2);
    int attempts = 0;
    while ((int)coins.size() < numCoins && attempts < 1000) {
        int x = dist(g), y = dist(g);
        if (maze[x][y] == 0 && find(coins.begin(), coins.end(), make_pair(x,y)) == coins.end() &&
            !(x == 1 && y == 1) && !(x == size-2 && y == size-1)) {
            coins.push_back({x, y});
        }
        attempts++;
    }
    return coins;
}

vector<pair<int,int>> findPath(const vector<vector<int>>& maze, pair<int,int> start, pair<int,int> end, int size) {
    vector<vector<bool>> visited(size, vector<bool>(size, false));
    vector<vector<pair<int,int>>> parent(size, vector<pair<int,int>>(size, {-1,-1}));
    queue<pair<int,int>> q;
    q.push(start);
    visited[start.first][start.second] = true;
    vector<pair<int,int>> dirs = {{0,1},{0,-1},{1,0},{-1,0}};
    while (!q.empty()) {
        auto cur = q.front(); q.pop();
        if (cur == end) {
            vector<pair<int,int>> path;
            while (cur != start) {
                path.push_back(cur);
                cur = parent[cur.first][cur.second];
            }
            path.push_back(start);
            reverse(path.begin(), path.end());
            return path;
        }
        for (auto& d : dirs) {
            int nx = cur.first + d.first, ny = cur.second + d.second;
            if (nx >= 0 && nx < size && ny >= 0 && ny < size && !visited[nx][ny] && maze[nx][ny] == 0) {
                visited[nx][ny] = true;
                parent[nx][ny] = cur;
                q.push({nx, ny});
            }
        }
    }
    return {};
}

void drawMaze(const vector<vector<int>>& maze, pair<int,int> player, const vector<pair<int,int>>& coins, int size, bool showPath, const vector<pair<int,int>>& path) {
    clearScreen();
    cout << colorize("┌" + string(size*2-1, '─') + "┐", GRAY) << endl;
    for (int i = 0; i < size; ++i) {
        string line = "│";
        for (int j = 0; j < size; ++j) {
            if (i == player.first && j == player.second) {
                line += colorize("@", GREEN);
            } else if (showPath && find(path.begin(), path.end(), make_pair(i,j)) != path.end()) {
                line += colorize(".", CYAN);
            } else if (find(coins.begin(), coins.end(), make_pair(i,j)) != coins.end()) {
                line += colorize("💎", YELLOW);
            } else if (maze[i][j] == 1) {
                line += colorize("█", GRAY);
            } else {
                line += " ";
            }
            if (j < size-1) line += " ";
        }
        line += "│";
        cout << line << endl;
    }
    cout << colorize("└" + string(size*2-1, '─') + "┘", GRAY) << endl;
}

char getch() {
    struct termios oldt, newt;
    char ch;
    tcgetattr(STDIN_FILENO, &oldt);
    newt = oldt;
    newt.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(STDIN_FILENO, TCSANOW, &newt);
    ch = getchar();
    tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
    return ch;
}

bool kbhit() {
    struct termios oldt, newt;
    int oldf;
    tcgetattr(STDIN_FILENO, &oldt);
    newt = oldt;
    newt.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(STDIN_FILENO, TCSANOW, &newt);
    oldf = fcntl(STDIN_FILENO, F_GETFL, 0);
    fcntl(STDIN_FILENO, F_SETFL, oldf | O_NONBLOCK);
    int ch = getchar();
    tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
    fcntl(STDIN_FILENO, F_SETFL, oldf);
    if (ch != EOF) {
        ungetc(ch, stdin);
        return true;
    }
    return false;
}

int main(int argc, char* argv[]) {
    int size = DEFAULT_SIZE;
    int coinsCount = DEFAULT_COINS;
    bool showPath = false;

    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "-s" && i+1 < argc) {
            size = stoi(argv[++i]);
        } else if (arg == "-c" && i+1 < argc) {
            coinsCount = stoi(argv[++i]);
        } else if (arg == "-p") {
            showPath = true;
        } else if (arg == "-h") {
            cout << "Usage: maze [options]\n  -s <N>   Size (default 10)\n  -c <N>   Coins (default 5)\n  -p       Show path\n";
            return 0;
        }
    }
    if (size % 2 == 0) size++;
    if (size < 5) size = 5;

    auto maze = generateMaze(size);
    auto coins = placeCoins(maze, coinsCount, size);
    pair<int,int> start = {1,1};
    pair<int,int> end = {size-2, size-1};
    pair<int,int> player = {1,1};
    int steps = 0;
    auto startTime = chrono::steady_clock::now();
    int collected = 0;
    auto records = loadRecords();
    string key = to_string(size) + "x" + to_string(size);

    vector<pair<int,int>> path;
    if (showPath) {
        path = findPath(maze, start, end, size);
    }

    while (true) {
        drawMaze(maze, player, coins, size, showPath, path);
        auto now = chrono::steady_clock::now();
        int elapsed = chrono::duration_cast<chrono::seconds>(now - startTime).count();
        cout << colorize("Шаги: " + to_string(steps) + "  Время: " + to_string(elapsed) + "с  Монеты: " + to_string(collected) + "/" + to_string(coins.size()), BOLD) << endl;
        cout << colorize("Управление: стрелки (WASD), Q - выход", BLUE) << endl;
        if (showPath && !path.empty()) cout << colorize("Кратчайший путь показан точками (cyan)", YELLOW) << endl;

        int dx = 0, dy = 0;
        if (kbhit()) {
            char ch = getch();
            if (ch == 'q' || ch == 'Q') {
                cout << colorize("Выход из игры.", YELLOW) << endl;
                return 0;
            }
            if (ch == 'w' || ch == 'W') dx = -1;
            else if (ch == 's' || ch == 'S') dx = 1;
            else if (ch == 'a' || ch == 'A') dy = -1;
            else if (ch == 'd' || ch == 'D') dy = 1;
            else if (ch == 27) { // стрелка
                char buf[2];
                if (getchar() == '[') {
                    char c = getchar();
                    if (c == 'A') dx = -1;
                    else if (c == 'B') dx = 1;
                    else if (c == 'D') dy = -1;
                    else if (c == 'C') dy = 1;
                }
            }
        }
        if (dx != 0 || dy != 0) {
            int nx = player.first + dx;
            int ny = player.second + dy;
            if (nx >= 0 && nx < size && ny >= 0 && ny < size && maze[nx][ny] == 0) {
                player = {nx, ny};
                steps++;
                auto it = find(coins.begin(), coins.end(), make_pair(nx, ny));
                if (it != coins.end()) {
                    coins.erase(it);
                    collected++;
                }
                if (player == end) {
                    elapsed = chrono::duration_cast<chrono::seconds>(chrono::steady_clock::now() - startTime).count();
                    cout << colorize("🎉 Поздравляем! Вы вышли за " + to_string(steps) + " шагов и " + to_string(elapsed) + " секунд!", GREEN) << endl;
                    cout << colorize("Собрано монет: " + to_string(collected), YELLOW) << endl;
                    int bonus = collected * 10;
                    cout << colorize("Бонус за монеты: +" + to_string(bonus) + " очков", YELLOW) << endl;
                    int total = steps + bonus;
                    cout << colorize("Итоговый счёт: " + to_string(total), BOLD) << endl;
                    if (!records.isMember(key) || steps < records[key]["steps"].asInt()) {
                        records[key]["steps"] = steps;
                        records[key]["time"] = elapsed;
                        records[key]["coins"] = collected;
                        saveRecords(records);
                        cout << colorize("🏆 Новый рекорд!", GREEN) << endl;
                    }
                    cout << colorize("Нажмите любую клавишу для выхода...", YELLOW) << endl;
                    getch();
                    return 0;
                }
            }
        }
        this_thread::sleep_for(chrono::milliseconds(30));
    }
    return 0;
}

File: maze.go
// maze.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"time"
	"golang.org/x/term"
)

const (
	DEFAULT_SIZE  = 10
	DEFAULT_COINS = 5
)

var (
	reset  = "\033[0m"
	red    = "\033[91m"
	green  = "\033[92m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	cyan   = "\033[96m"
	gray   = "\033[90m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

func clearScreen() {
	cmd := exec.Command("clear")
	if runtime.GOOS == "windows" {
		cmd = exec.Command("cmd", "/c", "cls")
	}
	cmd.Stdout = os.Stdout
	cmd.Run()
}

type Record struct {
	Steps int `json:"steps"`
	Time  int `json:"time"`
	Coins int `json:"coins"`
}

func loadRecords() map[string]Record {
	data, err := os.ReadFile(os.Getenv("HOME") + "/.maze_records.json")
	if err != nil {
		return make(map[string]Record)
	}
	var records map[string]Record
	json.Unmarshal(data, &records)
	return records
}

func saveRecords(records map[string]Record) {
	data, _ := json.MarshalIndent(records, "", "  ")
	os.WriteFile(os.Getenv("HOME")+"/.maze_records.json", data, 0644)
}

func generateMaze(size int) [][]int {
	maze := make([][]int, size)
	for i := range maze {
		maze[i] = make([]int, size)
		for j := range maze[i] {
			maze[i][j] = 1
		}
	}
	// Открываем проходы
	var dfs func(x, y int)
	dfs = func(x, y int) {
		maze[x][y] = 0
		dirs := [][2]int{{0, 2}, {0, -2}, {2, 0}, {-2, 0}}
		rand.Shuffle(len(dirs), func(i, j int) { dirs[i], dirs[j] = dirs[j], dirs[i] })
		for _, d := range dirs {
			nx, ny := x+d[0], y+d[1]
			if nx > 0 && nx < size-1 && ny > 0 && ny < size-1 && maze[nx][ny] == 1 {
				maze[x+d[0]/2][y+d[1]/2] = 0
				dfs(nx, ny)
			}
		}
	}
	dfs(1, 1)
	// Выход
	maze[size-2][size-1] = 0
	maze[size-1][size-1] = 0
	return maze
}

func placeCoins(maze [][]int, numCoins, size int) [][2]int {
	coins := [][2]int{}
	attempts := 0
	for len(coins) < numCoins && attempts < 1000 {
		x := rand.Intn(size-2) + 1
		y := rand.Intn(size-2) + 1
		if maze[x][y] == 0 && !contains(coins, [2]int{x, y}) && !(x == 1 && y == 1) && !(x == size-2 && y == size-1) {
			coins = append(coins, [2]int{x, y})
		}
		attempts++
	}
	return coins
}

func contains(slice [][2]int, item [2]int) bool {
	for _, v := range slice {
		if v == item {
			return true
		}
	}
	return false
}

func findPath(maze [][]int, start, end [2]int, size int) [][2]int {
	type Node struct {
		pos [2]int
		parent *Node
	}
	visited := make([][]bool, size)
	for i := range visited {
		visited[i] = make([]bool, size)
	}
	queue := []*Node{{pos: start, parent: nil}}
	visited[start[0]][start[1]] = true
	dirs := [][2]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]
		if cur.pos == end {
			// Восстанавливаем путь
			path := [][2]int{}
			for cur != nil {
				path = append([][2]int{cur.pos}, path...)
				cur = cur.parent
			}
			return path
		}
		for _, d := range dirs {
			nx, ny := cur.pos[0]+d[0], cur.pos[1]+d[1]
			if nx >= 0 && nx < size && ny >= 0 && ny < size && !visited[nx][ny] && maze[nx][ny] == 0 {
				visited[nx][ny] = true
				queue = append(queue, &Node{pos: [2]int{nx, ny}, parent: cur})
			}
		}
	}
	return nil
}

func drawMaze(maze [][]int, player [2]int, coins [][2]int, size int, showPath bool, path [][2]int) {
	clearScreen()
	// Верх
	fmt.Println(colorize("┌"+stringOf('─', size*2-1)+"┐", gray))
	for i := 0; i < size; i++ {
		line := "│"
		for j := 0; j < size; j++ {
			if (i == player[0] && j == player[1]) {
				line += colorize("@", green)
			} else if showPath && contains(path, [2]int{i, j}) {
				line += colorize(".", cyan)
			} else if contains(coins, [2]int{i, j}) {
				line += colorize("💎", yellow)
			} else if maze[i][j] == 1 {
				line += colorize("█", gray)
			} else {
				line += " "
			}
			if j < size-1 {
				line += " "
			}
		}
		line += "│"
		fmt.Println(line)
	}
	fmt.Println(colorize("└"+stringOf('─', size*2-1)+"┘", gray))
}

func stringOf(ch rune, n int) string {
	s := make([]rune, n)
	for i := range s {
		s[i] = ch
	}
	return string(s)
}

func main() {
	rand.Seed(time.Now().UnixNano())
	size := DEFAULT_SIZE
	coinsCount := DEFAULT_COINS
	showPath := false

	for i := 1; i < len(os.Args); i++ {
		if os.Args[i] == "-s" && i+1 < len(os.Args) {
			size, _ = strconv.Atoi(os.Args[i+1])
			i++
		} else if os.Args[i] == "-c" && i+1 < len(os.Args) {
			coinsCount, _ = strconv.Atoi(os.Args[i+1])
			i++
		} else if os.Args[i] == "-p" {
			showPath = true
		} else if os.Args[i] == "-h" {
			fmt.Println("Usage: maze [options]\n  -s <N>   Size (default 10)\n  -c <N>   Coins (default 5)\n  -p       Show path")
			return
		}
	}
	if size%2 == 0 {
		size++
	}
	if size < 5 {
		size = 5
	}

	maze := generateMaze(size)
	coins := placeCoins(maze, coinsCount, size)
	start := [2]int{1, 1}
	end := [2]int{size - 2, size - 1}
	player := [2]int{1, 1}
	steps := 0
	startTime := time.Now()
	collected := 0
	records := loadRecords()
	keyStr := fmt.Sprintf("%dx%d", size, size)

	var path [][2]int
	if showPath {
		path = findPath(maze, start, end, size)
	}

	oldState, err := term.MakeRaw(int(os.Stdin.Fd()))
	if err != nil {
		panic(err)
	}
	defer term.Restore(int(os.Stdin.Fd()), oldState)
	reader := bufio.NewReader(os.Stdin)

	for {
		drawMaze(maze, player, coins, size, showPath, path)
		elapsed := int(time.Since(startTime).Seconds())
		fmt.Println(colorize(fmt.Sprintf("Шаги: %d  Время: %ds  Монеты: %d/%d", steps, elapsed, collected, len(coins)), bold))
		fmt.Println(colorize("Управление: стрелки (WASD), Q - выход", blue))
		if showPath && path != nil {
			fmt.Println(colorize("Кратчайший путь показан точками (cyan)", yellow))
		}

		key, _, _ := reader.ReadRune()
		dx, dy := 0, 0
		switch key {
		case 'q', 'Q':
			fmt.Println(colorize("Выход из игры.", yellow))
			return
		case 'w', 'W':
			dx, dy = -1, 0
		case 's', 'S':
			dx, dy = 1, 0
		case 'a', 'A':
			dx, dy = 0, -1
		case 'd', 'D':
			dx, dy = 0, 1
		case '\x1b':
			// стрелка
			reader.ReadRune()
			dir, _, _ := reader.ReadRune()
			switch dir {
			case 'A':
				dx, dy = -1, 0
			case 'B':
				dx, dy = 1, 0
			case 'D':
				dx, dy = 0, -1
			case 'C':
				dx, dy = 0, 1
			}
		}
		if dx != 0 || dy != 0 {
			nx, ny := player[0]+dx, player[1]+dy
			if nx >= 0 && nx < size && ny >= 0 && ny < size && maze[nx][ny] == 0 {
				player[0], player[1] = nx, ny
				steps++
				// Проверка монет
				for i, c := range coins {
					if c[0] == nx && c[1] == ny {
						coins = append(coins[:i], coins[i+1:]...)
						collected++
						break
					}
				}
				if player == end {
					elapsed = int(time.Since(startTime).Seconds())
					fmt.Println(colorize(fmt.Sprintf("🎉 Поздравляем! Вы вышли за %d шагов и %d секунд!", steps, elapsed), green))
					fmt.Println(colorize(fmt.Sprintf("Собрано монет: %d", collected), yellow))
					bonus := collected * 10
					fmt.Println(colorize(fmt.Sprintf("Бонус за монеты: +%d очков", bonus), yellow))
					total := steps + bonus
					fmt.Println(colorize(fmt.Sprintf("Итоговый счёт: %d", total), bold))
					// Рекорды
					if rec, ok := records[keyStr]; !ok || steps < rec.Steps {
						records[keyStr] = Record{Steps: steps, Time: elapsed, Coins: collected}
						saveRecords(records)
						fmt.Println(colorize("🏆 Новый рекорд!", green))
					}
					fmt.Println(colorize("Нажмите любую клавишу для выхода...", yellow))
					reader.ReadRune()
					return
				}
			}
		}
	}
}
